src/shared/tpm2-util.c

```diff
@@ -5700,100 +5696,145 @@ int tpm2_unseal(Tpm2Context *c,
+                                        /* Due to bug #30546, if using RSA pubkey with the default exponent, we may
+                                         * need to set the exponent to the TPM special-case value of 0 and retry. */
+                                        log_debug("Policy hash mismatch, retrying with RSA pubkey exponent set to 0.");
+                                        pubkey_tpm2b.publicArea.parameters.rsaDetail.exponent = 0;
+                                        retry = true;
+                                        break;
+                                }
 #endif
                                 return log_debug_errno(SYNTHETIC_ERRNO(EPERM),
                                                        "Current policy digest does not match stored policy digest, cancelling "
                                                        "TPM2 authentication attempt.");
-                }
+                        }
+
+                        log_debug("Unsealing HMAC key for shard %zu.", shard);
+
+                        _cleanup_(Esys_Freep) TPM2B_SENSITIVE_DATA* unsealed = NULL;
+                        rc = sym_Esys_Unseal(
+                                        c->esys_context,
+                                        hmac_key->esys_handle,
+                                        policy_session->esys_handle,
+                                        encryption_session->esys_handle, /* use HMAC session to enable parameter encryption */
+                                        ESYS_TR_NONE,
+                                        &unsealed);
+                        if (rc == TPM2_RC_PCR_CHANGED && i > 0) {
+                                log_debug("A PCR value changed during the TPM2 policy session, restarting HMAC key unsealing (%u tries left).", i);
+                                retry = true;
+                                break;
+                        }
+                        if (rc != TPM2_RC_SUCCESS)
+                                return log_debug_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                                       "Failed to unseal HMAC key in TPM: %s", sym_Tss2_RC_Decode(rc));
 
-                log_debug("Unsealing HMAC key.");
+                        if (!iovec_append(&secret, &IOVEC_MAKE(unsealed->buffer, unsealed->size)))
+                                return log_oom_debug();
 
-                rc = sym_Esys_Unseal(
-                                c->esys_context,
-                                hmac_key->esys_handle,
-                                policy_session->esys_handle,
-                                encryption_session->esys_handle, /* use HMAC session to enable parameter encryption */
-                                ESYS_TR_NONE,
-                                &unsealed);
-                if (rc == TSS2_RC_SUCCESS)
+                        explicit_bzero_safe(unsealed->buffer, unsealed->size);
```

Comment:
Maybe, better to introduce a cleanup function that erases buffer then call `Esys_Free()`.
Otherwise, currently the buffer is not erased on some code path.